Improving Self-Hosting and Removing 3rd Party dependencies.

Podginator opened the pull request:
The intent of this PR is to improve the Self-Hosting documentation, to provide a working setup to get Omnivore working with Docker and Docker Compose. It intends to, as much as possible, remove third party dependencies and reliance on external infrastructure providers such as GCP. 

The aim is to establish feature parity, or near feature parity to the previously hosted service. This includes RSS support, webhook support, email newsletter, and PDF Support. 

The list of changes to date is below: 

- Create Dockerfile for Queue processing, which is used for parsing articles, alongside asynchronous tasks.
- Update and expose ImageProxy and use the latest version with ARM64 support.
- Create new docker-compose file in self-hosting/docker-compose.
- Provide a minimal .env file to be able to run the service using docker-compose.
- Created a guide for using [Cloudflare Tunnels](https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/) as a way to integrate with a device at your home. 
- Create a [NGINX](https://nginx.com) configuration for those looking to use NGINX Reverse Proxying for the service. 
- Replace use of Google Cloud Storage with [Minio](https://min.io/) an open-source layer compatible with the S3 API that can run on Device. 
    - This also allows other services, such as R2 and S3 to be the Storage Provider, if wanted.  
- Improvements to content-fetching to minimise instances where articles refused to parse. 
    - Also improved to not use puppeteer for some articles, instead relying on raw html. 
 - Overhaul the way email works, to ensure that there is an open source version. Three options are provided here. 
     - [Docker Mailserver](https://github.com/docker-mailserver/docker-mailserver): A production-ready fullstack but simple containerized mail server. This allows incoming emails to be received, parsed, and then added to Omnivore. 
     - [Amazon Simple Email Service](https://aws.amazon.com/ses/) A service provided by S3 that has a free tier. Allows for receiving of emails to a domain. Guide on how to set up in the Self-hosting readme. 
     - [Zapier](https://zapier.com/): Used as a way to integrate gmail to hosting. This can be realistically achieved using some of the gmail apis, also. 
- Replace [pspdfkit](https://www.nutrient.io/sd) - Which required a license and would display the following when using PDFS <img width="454" alt="image" src="https://github.com/user-attachments/assets/61e4061f-8740-4a09-9535-c59d5c1f798b" />
     
    - Have an option for the Native Browser PDF Viewer for PDF Files. This removes the highlight functionality, but is stable. 
    ![image](https://github.com/user-attachments/assets/70a3eb66-95df-49c8-9f15-20eecde12e8d)
    - Create a new pdf viewer using [PDF.js](https://mozilla.github.io/pdf.js/) an open source pdf library used as the backing for the PDF viewer in firefox. This option includes near feature parity (highlights, reading progress) with the pspdfkit, but may have some bugs. 
![image](https://github.com/user-attachments/assets/720cc2be-9168-4db7-84ec-ab05b436deab)
- Add some additional fixes to parsing articles, such as a Medium Parser, and a Wired parse
- Updated Docker images and software to the latest LTS version of Node (20.12) 

To-Do: 

- [x] Re-Enable Youtube features - such as extraction of Transcripts.
    - [x] Allow both an AI based feature for this, and a less formatted version.
- [ ] Provide a guide on how to get up and running user Kubernetes. 
- [ ] Provide a guide on how to get up and running with Tailscale. 
- [ ] Provide a guide on getting email to work with G-Mail without the use of an external server. 
- [ ] Attempt to provide a lighter-weight queuing system, and removal of Redis/Caching for single-user hosting.

samanthavbarron commented:
This is awesome, it looks like it's taking shape! I might try it out this weekend.

Are there contributions from the community that you can think of?

lovebes commented on packages/api/src/repository/storage/S3StorageClient.ts:
```diff
@@ -0,0 +1,140 @@
+import { getSignedUrl } from '@aws-sdk/s3-request-presigner'
+import type { Readable } from 'stream'
+
+// While this is listed as S3, for self hosting we will use MinIO, which is
```
Can we use Cloudflare R2 as well for self hosting? What was the decision behind using MinIO?

Podginator replied:
I'm not actually familiar with R2 - but anything that is S3 Compatible should work. Let me take a look later to see whether or not the Storage Client I built works with it. 

Minio was chosen because it can be self-hosted along with the rest of the application. There is a docker image, and it can all run on the same server without relying on anything external. 

I'm trying to ensure everything here can be run self-contained without any need for external services. 

That said, as with some of the email changes, I am looking into ways to simplify parts of it too, and having some external services is ok with me.

Mikilio replied:
To find suitable services, I recommend consulting r/self-hosted

volker-fr replied:
S3 is a nice idea, provides various options. How about local storage? This would reduce the required dependencies by one.

lovebes replied:
oh wow I didn't know Minio can be self-hosted! That sounds like a good idea.

Podginator replied:
> S3 is a nice idea, provides various options, including self hosted ones.
> 
> How about local storage? This would reduce the required dependencies by one.

The uploads are done via signed-urls, so while local-storage would be feasible it'd require a bit more development work.

jsifalda commented:
does it mean i would be able to deploy open-source omnivore to vercel? and be able to use this great app even after their shut-down? 🙏

RayBB commented:
If this gets worked out I'll add a template for easy self-hosting with Coolify

alexkotusenko commented:
Are you guys planning to add a docker container to self-host Omnivore?

diegonobre commented on self-hosting/GUIDE.md:
```diff
@@ -0,0 +1,369 @@
+### 3. Populate the .env file 
+
+There is a .env.example file located within the docker-compose folder that should give you the necessary environment variables to begin running. 
+You can use these by `mv .env.example.env`
```
Hello, playing with the guide I believe I just found a typo here. It's missing a space and should be `mv .env.example .env`

TransonQ commented:
Looking forward to a successful completion of this pr and thanks to the community for all their work.

thiswillbeyourgithub commented:
Is there also a plan to also rebuild the android apk to allow specifying a server url?

Podginator commented:
> Is there also a plan to also rebuild the android apk to allow specifying a server url?

You should be able to do this today. If you log out there's a self-hosting option on the android App.

Also the iOS App. 

I think there's a few bugs in both that need ironing out. In the android version you seemingly can't add links currently.

russtaylor commented on self-hosting/GUIDE.md:
```diff
@@ -0,0 +1,369 @@
+### 3. Populate the .env file 
+
+There is a .env.example file located within the docker-compose folder that should give you the necessary environment variables to begin running. 
+You can use these by `mv .env.example.env`
```
Just commenting w/that ^ as an official suggested change.

```suggestion
You can use these by `mv .env.example .env`
```

Podginator replied:
Hey - Thanks for this :)

archmagepi commented:
thanks to the community for all their work.

EnGassa commented:
@Podginator thanks for leading this effort. I'm sure this is really important to a lot of people in the community. How can we help push this along faster?

Gnosnay commented:
even i changed the URL, the login will still redirect to localhost.
anyway thx for your contribution @Podginator

Podginator commented:
> even i changed the URL, the login will still redirect to localhost. anyway thx for your contribution @Podginator

Have you changed it both in the env file, and in the docker compose file? For the web it needs to be changed too.

Gnosnay commented:
> > even i changed the URL, the login will still redirect to localhost. anyway thx for your contribution @Podginator
> 
> 
> 
> Have you changed it both in the env file, and in the docker compose file? For the web it needs to be changed too. 

Yeah I did. I guess I need to change code also?
I even rebuilt the images. Actually

submatrix commented on self-hosting/GUIDE.md:
```diff
@@ -0,0 +1,216 @@
+There is a .env.example file located within the docker-compose folder that should give you the necessary environment variables to begin running. 
+You can use these by `mv .env.example.env`
+
+The following environment variables should be changed to reflect where you are running your application. 
```
I believe the table below should also include `IMAGE_PROXY_URL` and `CLIENT_URL` as variables that need to be changed.

abinmn commented on self-hosting/docker-compose/docker-compose.yml:
```diff
@@ -0,0 +1,173 @@
+      - .env
+
+  content-fetch:
+    platform: linux/amd64
```
Is there any reason to specify platform? This is causing failure in ARM based machines

tlonic replied:
I've tested this on arm64 and can confirm this line is not necessary, content-fetch works fine on aarch64

mbaker3 commented on self-hosting/GUIDE.md:
```diff
@@ -0,0 +1,372 @@
+
+| Environment Variable             | Description                                    | Local Parameter         |
+|----------------------------------|------------------------------------------------|-------------------------|
+| BASE URL                         | The URL of the Front End of the Application.   | http://localhost:3000   |
```
Minor typo. Missing the `_` in `BASE_URL`

```suggestion
| BASE_URL                         | The URL of the Front End of the Application.   | http://localhost:3000   |
```

tlonic commented:
was able to get an instance up on an arm64 machine. as mentioned above content-fetch works fine on arm64. the only issue i had was imageproxy. in order to get it to work, i had to update the version to this one, which has arm64 binaries compiled: https://github.com/willnorris/imageproxy/pkgs/container/imageproxy/223956220?tag=main

everything else seemed to work fine, although i did not test the email stuff as I don't use that feature.

mbaker3 commented on self-hosting/nginx/nginx.conf:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    server {
+        listen 443;
+        server_name  omnivore.domain.com;
+
+
+        ssl_certification   /path/to/cert.crt;
+        ssl_certificate_key /path/to/cert.key;
+
+        ssl on;
```
Note sure if this is a version issue or not but on nginx 1.18.0
 - `ssl on` is deprecated in favour of putting it in the `listen` directive
 - `ssl_certification` is invalid and should be `ssl_certificate`

```suggestion
        listen 443 ssl;
        server_name  omnivore.domain.com;


        ssl_certificate   /path/to/cert.crt;
        ssl_certificate_key /path/to/cert.key;
```

Podginator replied:
Thanks for this. I tested it using a modified version with HTTP rather than HTTPS, hence why this was missed

Podginator commented on self-hosting/docker-compose/docker-compose.yml:
```diff
@@ -0,0 +1,172 @@
+    healthcheck:
+      test: ["CMD-SHELL", "nc -z 0.0.0.0 8080 || exit 1"]
+      interval: 15s
+      timeout: 90s
```
Certainly happy to increase this.

Mogelschrift commented:
> > > even i changed the URL, the login will still redirect to localhost. anyway thx for your contribution @Podginator
> > 
> > 
> > Have you changed it both in the env file, and in the docker compose file? For the web it needs to be changed too.
> 
> Yeah I did. I guess I need to change code also? I even rebuilt the images. Actually

I also face this issue ON my linux vps

Podginator commented:
> > > > even i changed the URL, the login will still redirect to localhost. anyway thx for your contribution @Podginator
> > > 
> > > 
> > > Have you changed it both in the env file, and in the docker compose file? For the web it needs to be changed too.
> > 
> > 
> > Yeah I did. I guess I need to change code also? I even rebuilt the images. Actually
> 
> I also face this issue ON my linux vps

There should be no need to change the code. I have recently updated the guide to include the IMAGE_PROXY_URL and CLIENT_URL parameters. Did you ensure that these were also changed? 

As well as the ones in the docker-compose file for the web. These are the things that write the redirect url, and so may be causing issues. 

I run mine on a Linux server, so that shouldn't be causing any issues.

Mogelschrift commented:
> > > > > even i changed the URL, the login will still redirect to localhost. anyway thx for your contribution @Podginator
> > > > 
> > > > 
> > > > Have you changed it both in the env file, and in the docker compose file? For the web it needs to be changed too.
> > > 
> > > 
> > > Yeah I did. I guess I need to change code also? I even rebuilt the images. Actually
> > 
> > 
> > I also face this issue ON my linux vps
> 
> There should be no need to change the code. I have recently updated the guide to include the IMAGE_PROXY_URL and CLIENT_URL parameters. Did you ensure that these were also changed?
> 
> As well as the ones in the docker-compose file for the web. These are the things that write the redirect url, and so may be causing issues.
> 
> I run mine on a Linux server, so that shouldn't be causing any issues.

I managed to get it running with a TLD and with rebuilding the whole images. But i still face a issue when pointing to https://read.mydomain.com/api where I get the following error: 

`Cannot GET /api/`

I made a NGINX custom location with the following parameters:

![image](https://github.com/user-attachments/assets/083aa83b-5cc7-49e1-bd1d-dba74e4b98b8)

submatrix commented:
> > > > > > even i changed the URL, the login will still redirect to localhost. anyway thx for your contribution @Podginator
> > > > > 
> > > > > 
> > > > > Have you changed it both in the env file, and in the docker compose file? For the web it needs to be changed too.
> > > > 
> > > > 
> > > > Yeah I did. I guess I need to change code also? I even rebuilt the images. Actually
> > > 
> > > 
> > > I also face this issue ON my linux vps
> > 
> > 
> > There should be no need to change the code. I have recently updated the guide to include the IMAGE_PROXY_URL and CLIENT_URL parameters. Did you ensure that these were also changed?
> > As well as the ones in the docker-compose file for the web. These are the things that write the redirect url, and so may be causing issues.
> > I run mine on a Linux server, so that shouldn't be causing any issues.
> 
> I managed to get it running with a TLD and with rebuilding the whole images. But i still face a issue when pointing to https://read.mydomain.com/api where I get the following error:
> 
> `Cannot GET /api/`
> 
> I made a NGINX custom location with the following parameters:
> 
> ![image](https://private-user-images.githubusercontent.com/190207139/391327821-083aa83b-5cc7-49e1-bd1d-dba74e4b98b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMwNjcwNjUsIm5iZiI6MTczMzA2Njc2NSwicGF0aCI6Ii8xOTAyMDcxMzkvMzkxMzI3ODIxLTA4M2FhODNiLTVjYzctNDllMS1iZDFkLWRiYTc0ZTRiOThiOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMVQxNTI2MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZjdiYjc3MGFiZmI2Zjc0ZTU2N2QxMjg2ZDY5MTEyMjYwZTMxYWM2MDU2NjhmMTNjMDE3NDhlMjMxYjhhYjY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.aN9NNeVVqQqmNPUVYVy6Xupw84YUNyz7mNvH0s2NJCA)

Is it because you've pointed it to https but your nginx config specifies http as per your screenshot?

submatrix commented on self-hosting/GUIDE.md:
```diff
@@ -0,0 +1,375 @@
+
+    server {
+        listen 80;
+        return 301 https://$host$request_uri
```
Could you sync this example .conf with your updated version in nginx/nginx.conf? For example, you added a semicolon on this line to fix a bug but the readme still has the bug present.

mbaker3 replied:
There are quite a few changes. It's probably easier to just link to the file itself vs trying to keep this up to date.

Mogelschrift commented:
> > > > > > > even i changed the URL, the login will still redirect to localhost. anyway thx for your contribution @Podginator
> > > > > > 
> > > > > > 
> > > > > > Have you changed it both in the env file, and in the docker compose file? For the web it needs to be changed too.
> > > > > 
> > > > > 
> > > > > Yeah I did. I guess I need to change code also? I even rebuilt the images. Actually
> > > > 
> > > > 
> > > > I also face this issue ON my linux vps
> > > 
> > > 
> > > There should be no need to change the code. I have recently updated the guide to include the IMAGE_PROXY_URL and CLIENT_URL parameters. Did you ensure that these were also changed?
> > > As well as the ones in the docker-compose file for the web. These are the things that write the redirect url, and so may be causing issues.
> > > I run mine on a Linux server, so that shouldn't be causing any issues.
> > 
> > 
> > I managed to get it running with a TLD and with rebuilding the whole images. But i still face a issue when pointing to https://read.mydomain.com/api where I get the following error:
> > `Cannot GET /api/`
> > I made a NGINX custom location with the following parameters:
> > ![image](https://private-user-images.githubusercontent.com/190207139/391327821-083aa83b-5cc7-49e1-bd1d-dba74e4b98b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMwNjcwNjUsIm5iZiI6MTczMzA2Njc2NSwicGF0aCI6Ii8xOTAyMDcxMzkvMzkxMzI3ODIxLTA4M2FhODNiLTVjYzctNDllMS1iZDFkLWRiYTc0ZTRiOThiOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMVQxNTI2MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZjdiYjc3MGFiZmI2Zjc0ZTU2N2QxMjg2ZDY5MTEyMjYwZTMxYWM2MDU2NjhmMTNjMDE3NDhlMjMxYjhhYjY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.aN9NNeVVqQqmNPUVYVy6Xupw84YUNyz7mNvH0s2NJCA)
> 
> Is it because you've pointed it to https but your nginx config specifies http as per your screenshot?

With the nginx config I point per http to the specific api docker container over https. If I change it to http i get no connection at all. At the moment my login and the content fetch are working. The content has no images and the URL Save over read.mydomain.com/api/save?={url} are not working. So I assume that the connection to the image-proxy and the api container are not working.

mbaker3 commented on self-hosting/GUIDE.md:
````diff
@@ -0,0 +1,375 @@
+        }
+    }
+}
+```
````
Easier than keeping the config up to date with the example file.

```suggestion
[See here](nginx/nginx.conf)
```

Mogelschrift commented:
> > > > > > > > > even i changed the URL, the login will still redirect to localhost. anyway thx for your contribution @Podginator
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Have you changed it both in the env file, and in the docker compose file? For the web it needs to be changed too.
> > > > > > > 
> > > > > > > 
> > > > > > > Yeah I did. I guess I need to change code also? I even rebuilt the images. Actually
> > > > > > 
> > > > > > 
> > > > > > I also face this issue ON my linux vps
> > > > > 
> > > > > 
> > > > > There should be no need to change the code. I have recently updated the guide to include the IMAGE_PROXY_URL and CLIENT_URL parameters. Did you ensure that these were also changed?
> > > > > As well as the ones in the docker-compose file for the web. These are the things that write the redirect url, and so may be causing issues.
> > > > > I run mine on a Linux server, so that shouldn't be causing any issues.
> > > > 
> > > > 
> > > > I managed to get it running with a TLD and with rebuilding the whole images. But i still face a issue when pointing to https://read.mydomain.com/api where I get the following error:
> > > > `Cannot GET /api/`
> > > > I made a NGINX custom location with the following parameters:
> > > > ![image](https://private-user-images.githubusercontent.com/190207139/391327821-083aa83b-5cc7-49e1-bd1d-dba74e4b98b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMwNjcwNjUsIm5iZiI6MTczMzA2Njc2NSwicGF0aCI6Ii8xOTAyMDcxMzkvMzkxMzI3ODIxLTA4M2FhODNiLTVjYzctNDllMS1iZDFkLWRiYTc0ZTRiOThiOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMVQxNTI2MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZjdiYjc3MGFiZmI2Zjc0ZTU2N2QxMjg2ZDY5MTEyMjYwZTMxYWM2MDU2NjhmMTNjMDE3NDhlMjMxYjhhYjY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.aN9NNeVVqQqmNPUVYVy6Xupw84YUNyz7mNvH0s2NJCA)
> > > 
> > > 
> > > Is it because you've pointed it to https but your nginx config specifies http as per your screenshot?
> > 
> > 
> > With the nginx config I point per http to the specific api docker container over https. If I change it to http i get no connection at all. At the moment my login and the content fetch are working. The content has no images and the URL Save over read.mydomain.com/api/save?={url} are not working. So I assume that the connection to the image-proxy and the api container are not working.
> 
> Will preface this by saying that I changed the NGINX to run with port 80 rather than ssl.
> 
> The cannot get /api/ makes sense. There is no route /api - The fact that you're getting that express error shows that the API is indeed working. Going there myself I also get the `Cannot GET /api` - but the rest of my application works fine.
> 
> When changing the URL, are you including /api in the places where you are giving the API address. This will cause the url to become /api/api/graphql rather than /api.
> 
> If you want to continue using that, you can add the following in the NGINX config, which will convert it back to a single /api.
> 
> ```
> location /api {
>   rewrite ^/api/api/(.*)$ /api/$1 break;
>   proxy_pass http://omnivore_backend;
> }
> ```
> 
> As for imageproxy - I'm not sure what's happening there. For me, this just worked off the bat, but I did have to include /images in the Environment Variable `IMAGE_PROXY_URL=http://localhost/images`
> 
> If you are using subdomains instead you can modify the nginx to include this. In the mean time I would just use this.
> 
> Feel free to reach out on Discord too, we can try further debugging there.

Thanks for the tip with the environment variable, images work now :)  With the API Issue I am trying to use a browser extension called "Omnivore List Popup". The standard URL for the API in the settings is: 

https://api-prod.omnivore.app/api/graphql

I tried using the following URL:

https://read.mydomain.com/api/graphql

which is not working. Any idea why?
For the other issue I faced (the Cannot get /api/) I tried to save pages via the bookmarklet:

`javascript:void(open('https://read.mydomain.com/api/save?url='+encodeURIComponent(location.href),'Omnivore'))`

Still not working. I need to say, that I didnt have the /api/api issue u mentioned in the last comment.

Mogelschrift commented:
> > > > > > > > > > > even i changed the URL, the login will still redirect to localhost. anyway thx for your contribution @Podginator
> > > > > > > > > > 
> > > > > > > > > > 
> > > > > > > > > > Have you changed it both in the env file, and in the docker compose file? For the web it needs to be changed too.
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > Yeah I did. I guess I need to change code also? I even rebuilt the images. Actually
> > > > > > > > 
> > > > > > > > 
> > > > > > > > I also face this issue ON my linux vps
> > > > > > > 
> > > > > > > 
> > > > > > > There should be no need to change the code. I have recently updated the guide to include the IMAGE_PROXY_URL and CLIENT_URL parameters. Did you ensure that these were also changed?
> > > > > > > As well as the ones in the docker-compose file for the web. These are the things that write the redirect url, and so may be causing issues.
> > > > > > > I run mine on a Linux server, so that shouldn't be causing any issues.
> > > > > > 
> > > > > > 
> > > > > > I managed to get it running with a TLD and with rebuilding the whole images. But i still face a issue when pointing to https://read.mydomain.com/api where I get the following error:
> > > > > > `Cannot GET /api/`
> > > > > > I made a NGINX custom location with the following parameters:
> > > > > > ![image](https://private-user-images.githubusercontent.com/190207139/391327821-083aa83b-5cc7-49e1-bd1d-dba74e4b98b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMwNjcwNjUsIm5iZiI6MTczMzA2Njc2NSwicGF0aCI6Ii8xOTAyMDcxMzkvMzkxMzI3ODIxLTA4M2FhODNiLTVjYzctNDllMS1iZDFkLWRiYTc0ZTRiOThiOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMVQxNTI2MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZjdiYjc3MGFiZmI2Zjc0ZTU2N2QxMjg2ZDY5MTEyMjYwZTMxYWM2MDU2NjhmMTNjMDE3NDhlMjMxYjhhYjY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.aN9NNeVVqQqmNPUVYVy6Xupw84YUNyz7mNvH0s2NJCA)
> > > > > 
> > > > > 
> > > > > Is it because you've pointed it to https but your nginx config specifies http as per your screenshot?
> > > > 
> > > > 
> > > > With the nginx config I point per http to the specific api docker container over https. If I change it to http i get no connection at all. At the moment my login and the content fetch are working. The content has no images and the URL Save over read.mydomain.com/api/save?={url} are not working. So I assume that the connection to the image-proxy and the api container are not working.
> > > 
> > > 
> > > Will preface this by saying that I changed the NGINX to run with port 80 rather than ssl.
> > > The cannot get /api/ makes sense. There is no route /api - The fact that you're getting that express error shows that the API is indeed working. Going there myself I also get the `Cannot GET /api` - but the rest of my application works fine.
> > > When changing the URL, are you including /api in the places where you are giving the API address. This will cause the url to become /api/api/graphql rather than /api.
> > > If you want to continue using that, you can add the following in the NGINX config, which will convert it back to a single /api.
> > > ```
> > > location /api {
> > >   rewrite ^/api/api/(.*)$ /api/$1 break;
> > >   proxy_pass http://omnivore_backend;
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > As for imageproxy - I'm not sure what's happening there. For me, this just worked off the bat, but I did have to include /images in the Environment Variable `IMAGE_PROXY_URL=http://localhost/images`
> > > If you are using subdomains instead you can modify the nginx to include this. In the mean time I would just use this.
> > > Feel free to reach out on Discord too, we can try further debugging there.
> > 
> > 
> > Thanks for the tip with the environment variable, images work now :) With the API Issue I am trying to use a browser extension called "Omnivore List Popup". The standard URL for the API in the settings is:
> > https://api-prod.omnivore.app/api/graphql
> > I tried using the following URL:
> > https://read.mydomain.com/api/graphql
> > which is not working. Any idea why? For the other issue I faced (the Cannot get /api/) I tried to save pages via the bookmarklet:
> > `javascript:void(open('https://read.mydomain.com/api/save?url='+encodeURIComponent(location.href),'Omnivore'))`
> > Still not working. I need to say, that I didnt have the /api/api issue u mentioned in the last comment.
> 
> I've figured out the extension issue.
> 
> The extension has the following in it's manifest to allow cross-site calls. <img alt="image" width="712" src="https://private-user-images.githubusercontent.com/3532775/391883971-143c3757-9dba-445e-b001-b965d524ae4d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyMTczNzEsIm5iZiI6MTczMzIxNzA3MSwicGF0aCI6Ii8zNTMyNzc1LzM5MTg4Mzk3MS0xNDNjMzc1Ny05ZGJhLTQ0NWUtYjAwMS1iOTY1ZDUyNGFlNGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDNUMDkxMTExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzBlMTZkYTFjNzBlM2NlZmVkZmYxYTlkNzNhNmQ3MjBhZmI1MGYyYWJhMTlkMmFmZGE5NGZlZTM4MGIwOGEzNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.PzeuQ-lr04rpDKr1MSy2BmWOSGhpZISQO_ThhTmDtyQ">
> 
> When changing the URL to yours there will be a CORS issue when trying to call it.
> 
> To fix this you can do the following:
> 
> Either download the source code here, and in the manifest file add your URL to the array found here. https://github.com/herrherrmann/omnivore-list-popup/blob/27d20f951642ccb8c0f578d9ab05681878470af0/src/manifest.chrome.json#L18
> 
> Alternatively you can download the extension using something like https://chromewebstore.google.com/detail/chrome-extension-source-v/jifpbeccnghkjeaalbbjmodiffmgedin and modify the manifest yourself.
> 
> After installing the [unpacked extension](https://knowledge.workspace.google.com/kb/load-unpacked-extensions-000005962) you will have your contents load. <img alt="image" width="349" src="https://private-user-images.githubusercontent.com/3532775/391885553-1709ecba-20cb-4680-8dc9-8a32b0ab981f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyMTczNzEsIm5iZiI6MTczMzIxNzA3MSwicGF0aCI6Ii8zNTMyNzc1LzM5MTg4NTU1My0xNzA5ZWNiYS0yMGNiLTQ2ODAtOGRjOS04YTMyYjBhYjk4MWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDNUMDkxMTExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2E4YzkyNThlMDFlMTBhZjRmMjEwNWMyMzdkYzM5ZjIyOTg5YTk4YTQ0MDZkMzNmZWY4YTU5ZjkzNTJkZmMyNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.XMYHaBm8614BciNiVHlNl0q4GWjuYBT5ZJZwsBAn5bE">
> 
> As for the other problem, I think I have an idea of how to fix this - but I am unable to test right now.
> 
> In your NGINX add the following:
> 
> ```
> location /api/save {
>             proxy_pass http://omnivore_web;
>         }
> ```
> 
> Then reload.
> 
> I think the problem might be that this is a route for the frontend application, but it's directing it to the backend. You can see in the past we had a similar problem here:
> 
> ```
> location /api/client/auth {
>             proxy_pass http://omnivore_web;
> }
> ```
> 
> Give that a try and let me know. If it's solved with that I will add it to the official one.

I couldn't try the extension atm but the solution with the nginx location worked. Saving by URL is now working properly!

git2wing commented:
It has been confirmed that for the Android App clipping function, even if the self-hosted address is configured for Add and Share Link, only the official Omnivore address will be used to connect.

Search for "ApolloClient.Builder" in the app code , and refer to "authenticatedApolloClient() = ApolloClient.Builder().serverUrl(serverUrl))" in networker.kt , to correct these codes. Rebuild and package apk。then with the self-hosted service, the Android app will be able to add and share link.

thiswillbeyourgithub commented:
Hi everyone, very happy to see the level of engagement here! unfortunately this is all quite technical so I was wondering if there was some readme or roadmap somewhere that listed the crurent feasibility of self hosting, gotchas, maturity etc as well as how to import the data we exported from the original server. I surely can't be the only one to have the skills to run a docker ready system but lack the knowledge to follow closely this (very active) PR. And thank you all very much!

submatrix commented:
> Hi everyone, very happy to see the level of engagement here! unfortunately this is all quite technical so I was wondering if there was some readme or roadmap somewhere that listed the crurent feasibility of self hosting, gotchas, maturity etc as well as how to import the data we exported from the original server. I surely can't be the only one to have the skills to run a docker ready system but lack the knowledge to follow closely this (very active) PR. And thank you all very much!

Your best bet is probably to join us in the Omnivore discord, specifically the #self-hosting channel. We have some guides there.

luca-git commented:
Will this ever happen?

Podginator commented:
@luca-git I haven't got the permission to merge, but this branch is a perfectly viable way of getting the application up and running as is. I'll speak with @jacksonh about when the merge should take place.
